### PR TITLE
Bindings

### PR DESCRIPTION
**Pull request by MarkusKonk**
First version of the bindings API with one example.

### CODE REVIEW

File: docs/compendium/bindings.md [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+  "numerical_result": "3.14",
+  "dataset": [{
+    "dataset_file": ["data.csv"],
+    "dataset_columns" : ["measured_values"],
+    "dataset_rows": ["1-100"]
+   }]  
+}
+```
+
+### Request body properties
```
Comment (nuest) [inline docs/compendium/bindings.md]: Would be helpful to define the data format of the properties. `result_lines` could be a JSON array of numbers, or a string depicting a line range, or... possibly just link to the line format definition somewhere else?

File: docs/compendium/bindings.md [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+
+- `id` - id of the compendium
+- `purpose` - task that an author would like to provide
+- `main_file` - main RMarkdown file used to compute the display file
+- `result_line` - line(s) where the result occurs
```
Comment (nuest) [inline docs/compendium/bindings.md]: Don't understand this property's meaning, also inconsistent with other properties who are plural.

Is this about the code lines where the result is "created", i.e. plotted or saved to a file? Or is it about the lines where the result is actually shown, e.g. if it is a number?

File: docs/compendium/bindings.md [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+  - `dataset_columns` - array of columns used to compute the number
+  - `dataset_rows` - array of row ranges used to compute the number
+
+### Response
+
+```json
+204 OK
+```
+
+### Error responses for bindings creation
```
Comment (nuest) [inline docs/compendium/bindings.md]: Do you plan any validation in the API, e.g. checking if the line ranges are actually matching the file, etc.

File: docs/compendium/bindings.md [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+8. Manipulate a parameter used to compute a figure.
+9. Select data subsets used to produce a figure.
+10. Select data subsets used to produce a specific numerical result.
+
+## Create a new binding
+
+When creating a new _Executable Research Compendium_ (ERC), authors will reach the "Create bindings section".
+Each purpose has own needs.
+
+`POST /api/v1/binding`
```
Comment (nuest) [inline docs/compendium/bindings.md]: How can a user `GET` the existing bindings?
We should mention here where they are stored within the compendium documents.

File: docs/compendium/bindings.md [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+5. Show the source code underlying a specific numerical result.
+6. Show the source code underlying a figure.
+7. Manipulate a parameter used to compute a specific numerical result.
+8. Manipulate a parameter used to compute a figure.
+9. Select data subsets used to produce a figure.
+10. Select data subsets used to produce a specific numerical result.
+
+## Create a new binding
+
+When creating a new compendium, authors will reach the "Create bindings section".
```
Comment (nuest) [inline docs/compendium/bindings.md]: This is the API documentation, the "Create bindings section" is part of the reference user interface.

What are the prerequisites for a binding from an API perspective? Does the compendium have to be a candidate, or not?

File: docs/compendium/bindings.md [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+Each purpose has own needs.
```
Comment (nuest) [inline docs/compendium/bindings.md]: Purposes are already described above. What are these "own needs"? There is only one example below.

File: docs/compendium/bindings.md [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+
+`POST /api/v1/binding`
+
+### Request body for a new binding depending on the purpose:
+
+3. Show the dataset underlying a specific numerical result.
+```json
+{
+  "id":"compendium_id",
+  "purpose": "showNumResultData",
```
Comment (nuest) [inline docs/compendium/bindings.md]: There seems to be an identifier for the purpose, suggest to add the technical identifier for each purpose to the list of purposes.

File: docs/compendium/bindings.md (hunk 5 above)
Comment (MarkusKonk) [reply docs/compendium/bindings.md]: "Create bindings section" is not about explaining how it works but which info is needed - reworded

Included prerequisites

File: docs/compendium/bindings.md (hunk 6 above)
Comment (MarkusKonk) [reply docs/compendium/bindings.md]: Reworded

File: docs/compendium/bindings.md [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+  "main_file": "main_file.Rmd",
+  "result_line": "19-20",
+  "numerical_result": "3.14",
+  "dataset": [{
+    "dataset_file": ["data.csv"],
```
[the remark this answers is not in the document]
Comment (MarkusKonk) [reply docs/compendium/bindings.md]: added note at the beginning

File: docs/compendium/bindings.md (hunk 1 above)
Comment (MarkusKonk) [reply docs/compendium/bindings.md]: included data formats

File: docs/compendium/bindings.md (hunk 2 above)
Comment (MarkusKonk) [reply docs/compendium/bindings.md]: reworded

File: docs/compendium/bindings.md (hunk 3 above)
Comment (MarkusKonk) [reply docs/compendium/bindings.md]: integrated a few

**Comment by nuest** [comment]
There's some minor glitches (like invalid JSON in the examples), but I understand what this endpoint should do and how. I suggest to continue with the implementation and keep this PR updated until the server-side implementation is done - or is there no need for implementation @MarkusKonk ?

**Comment by MarkusKonk** [comment]
Ok. Thanks. Yes, there will be some implementation of a microservice. I will tell you about that tomorrow.

**Comment by nuest** [comment]
@MarkusKonk Is this up to date?

**Pull request closed without merging**